akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/MessageAdapterSpec.scala

```diff
@@ -0,0 +1,235 @@
+      case class Pong2(greeting: String) extends Response
+
+      case class Wrapped(qualifier: String, response: Response)
+
+      val pingPong = spawn(Behaviors.immutable[Ping] { (_, msg) ⇒
+        msg match {
+          case Ping1(sender) ⇒
+            sender ! Pong1("hello-1")
+            Behaviors.same
+          case Ping2(sender) ⇒
+            // doing something terribly wrong
+            sender ! Pong2("hello-2")
+            Behaviors.same
+        }
+      })
+
+      val probe = TestProbe[Wrapped]()
+
+      val snitch = Behaviors.deferred[Wrapped] { (ctx) ⇒
+
+        val replyTo1 = ctx.messageAdapter[Pong1](pong ⇒ Wrapped(qualifier = "1", pong))
+        pingPong ! Ping1(replyTo1)
+        // doing something terribly wrong
+        // Pong2 message adapter not registered
+        pingPong ! Ping2(replyTo1.asInstanceOf[ActorRef[Pong2]])
+        pingPong ! Ping1(replyTo1)
+
+        Behaviors.immutable {
+          case (_, wrapped) ⇒
+            probe.ref ! wrapped
+            Behaviors.same
+        }
+      }
+
+      EventFilter.warning(start = "unhandled message", occurrences = 1).intercept {
+        spawn(snitch)
+      }
+
+      probe.expectMsg(Wrapped("1", Pong1("hello-1")))
+      // hello-2 discarded because it was wrong type
+      probe.expectMsg(Wrapped("1", Pong1("hello-1")))
+    }
+
+    "stop when exception from adapter" in {
+      case class Ping(sender: ActorRef[Pong])
+      case class Pong(greeting: String)
+      case class Wrapped(count: Int, response: Pong)
+
+      val pingPong = spawn(Behaviors.immutable[Ping] { (_, ping) ⇒
+        ping.sender ! Pong("hello")
+        Behaviors.same
+      })
+
+      val probe = TestProbe[Any]()
+
+      val snitch = Behaviors.deferred[Wrapped] { (ctx) ⇒
+
+        var count = 0
+        val replyTo = ctx.messageAdapter[Pong] { pong ⇒
+          count += 1
+          if (count == 3) throw new TestException("boom")
+          else Wrapped(count, pong)
+        }
+        (1 to 4).foreach { _ ⇒
+          pingPong ! Ping(replyTo)
+        }
+
+        Behaviors.immutable[Wrapped] {
+          case (_, wrapped) ⇒
+            probe.ref ! wrapped
+            Behaviors.same
+        }.onSignal {
+          case (_, PostStop) ⇒
+            probe.ref ! "stopped"
+            Behaviors.same
+        }
+      }
+
+      EventFilter.warning(pattern = ".*received dead letter.*", occurrences = 1).intercept {
+        EventFilter[TestException](occurrences = 1).intercept {
+          spawn(snitch)
+        }
+      }
+
+      probe.expectMsg(Wrapped(1, Pong("hello")))
+      probe.expectMsg(Wrapped(2, Pong("hello")))
+      // exception was thrown for  3
+
+      // FIXME One thing to be aware of is that the supervision strategy of the Behavior is not
+      // used for exceptions from adapters. Should we instead catch, log, unhandled, and resume?
+      // It's kind of "before" the message arrives.
+      probe.expectMsg("stopped")
```

Comment:
@johanandren What do you think about this? Would be same for `ask`. I think we should catch such exceptions and not involve supervision, with the motivation that it's "before" receiving the actual message.

Reply:
What about wrapping in a special `AdapterException` and use that for both `ask` and here?

Reply:
That misses my point. It would be very difficult to use the supervision, since that is in the Behavior, and such exception occur before we call the Behavior.

The reason it is stopping currently is because that is the default for typed when a typed actor (`ActorAdapter`) throws, and it wouldn't change if someone "installed" a supervision strategy for such exception.

Reply:
aaaah, yes, I completely missed that. Getting an unexpected message back is somewhat serious though, isn't logging error and stopping the right thing to do?